### ISSUE TITLE
Incorrect Category

### ISSUE THREAD

**Issue by KosDrozd** [nunit/docs#251]

**Comment by mikkelbu** [comment]
Hi @KosDrozd. Can you please provide some more information than just "Incorrect Category"?

**Comment by KosDrozd** [comment]
Given I have one Test with different category

[Test, Author("kostiantyn.drozd")]
        [Category("cat_Example1")]
        [Category("cat_Example2")]
        [Category("cat_Example3")]
        [Category("cat_Example4")]
        [Category("cat_Example5")]
        [Category("cat_Example6")]
        [Category("cat_Example7")]
        public void MyTest()
        {
            string category = TestContext.CurrentContext.Test.Properties.Get("Category").ToString();
        }

When I run my tests for category cat_Example5
nunit3-console mytest.dll --where "cat == cat_Example5"

Then
string category = "cat_Example1"
**I expected that string category = "cat_Example5"**

I am not sure that this is bug, but this is very strange

**Comment by KosDrozd** [comment]
@mikkelbu  Sorry, this is my firs issue

**Comment by mikkelbu** [comment]
@KosDrozd It is completely ok (I also noticed that you just created your github account today. Welcome 😄 ). I think that you are missing an assertion in the test, but I also think I get the gist of it.

The `Get` property only returns the first element (if there are any), see e.g. https://github.com/nunit/nunit/blob/acdc5477eae0aa489331199ec07053c9e43b69f2/src/NUnitFramework/framework/Internal/PropertyBag.cs#L85. If you instead use the indexer then you'll get all. Something like
```
IList  categories = TestContext.CurrentContext.Test.Properties["Category"];
```
where each element will be a string,

**Comment by KosDrozd** [comment]
Thanks @mikkelbu, but I need take the current category.
Could you give me some information about any workaround?
How can I take cat_Example5 (category what running now) ?

**Comment by CharliePoole** [comment]
The code that @mikkelbu gave you will return all the categories on the test. The test doesn't have any knowledge of __why__ it is being run. It could be that you are running all tests or you selected it by name,or you selected it by category for example.

This is part of the overall architecture of NUnit tests. Tests do not know who is running them, why they are being run etc. The intention is that a test should always work the same way, no matter how it got run.

If you can tell us __why__ you want the information - what you would do with it - we can probably suggest some workarounds.

**Comment by KosDrozd** [comment]
In category I set Build_Number, and the implementation of some steps depends on the build number.

[Test, Author("kostiantyn.drozd")]
[Category("build_1")]
[Category("build_2")]
public void MyTest()
{
string currentCategory = TestContext.CurrentContext.Test.Properties.Get("Category").ToString();
SomeMethod(currentCategory);
}

P.S. I thought about the separation of such tests, but 95% of the steps are not different, and I do not want to duplicate the code.

Also I try use TestCase Attribute with build_1 as input parameters, but I dont now how I can run this test by build_1

**Comment by CharliePoole** [comment]
That's a use if Category that the original designers considered to be an anti-pattern and took steps to make it impossible.

However, it sounds as if you are trying to accomplish what is already available through run parameters. Have you looked at that feature? See TextContext.

**Comment by KosDrozd** [comment]
@CharliePoole,
Yes I can take currentCategor from run parameters, but I can not run test on nunit3-console by categories.

[TestCase("build_1"), Author("kostiantyn.drozd"), Category("build_1")]
        [TestCase("build_2"), Author("kostiantyn.drozd"), Category("build_2")]
        public void MyTest(string currentCategor)
        {
            var x= TestContext.CurrentContext.Test.Arguments;
            File.AddLineToFile("D:\\3.txt", currentCategor);
        }
When I run
nunit3-console.exe TEST.dll --where "cat == build_1"
I **expected that only one test will runed** ([TestCase("build_1"), Author("kostiantyn.drozd"), Category("build_1")])
**Actual result: Run 2 test** 
[TestCase("build_1"), Author("kostiantyn.drozd"), Category("build_1")]
and
[TestCase("build_2"), Author("kostiantyn.drozd"), Category("build_2")]

**Comment by jnm2** [comment]
I agree that this is not what categories are for. You should either stop using categories and use `TestContext.Parameters`, or split the test into multiple test case or test methods with one category each.

I don't think you realize this but this is a C# rule, not an NUnit rule. Your code:

```cs
[TestCase("build_1"), Author("kostiantyn.drozd"), Category("build_1")]
[TestCase("build_2"), Author("kostiantyn.drozd"), Category("build_2")]
```

Is no different in C# than this code:
```cs
[TestCase("build_1")]
[TestCase("build_2")]
[Author("kostiantyn.drozd")]
[Author("kostiantyn.drozd")]
[Category("build_1")]
[Category("build_2")]
```

I think what you meant to use is the Category property of the TestCase attribute:
```cs
[Author("kostiantyn.drozd")]
[TestCase("build_1", Category = "build_1")]
[TestCase("build_2", Category = "build_2")]
```

Perhaps jsut fixing that would solve your problem?

**Comment by jnm2** [comment]
I'm glad it helped and don't be afraid to ask about anything. If something feels strange to you, I think it's often an indication that you would benefit from understanding it better. 😃

**Comment by CharliePoole** [comment]
@KosDrozd The correction pointed out by @jnm2 will run the proper tests. It has the drawback of forcing you to specify the build twice.

That's because you have misunderstood what are "Run Parameters." They are not the same as test case parameters but are general parameters available to every test in your run that needs them.

You specify a run parameter from the console using the `--params` option, e.g.: `--params Build=build1` 

You access this parameter in your test using `TestContext.TestParameters["Build"]`

**Comment by KosDrozd** [comment]
So much better.
Thank you so much
